### ISSUE TITLE
FREQ Add Craft 3.7 Nested Matrix field GraphQL mutation support

### ISSUE THREAD

**Issue by humandevs** [engram-design/SuperTable#425]
**What are you trying to do?**
Mutate nested Matrix fields.

**What's your proposed solution?**
Craft 3.7 now supports mutations for nested Matrix fields inside a Super Table field. See the comment on July 13th regarding how to add this:
https://github.com/craftcms/cms/issues/9139

**Additional context**
Original Super-Table bug report here: https://github.com/verbb/super-table/issues/416

"When adding support for Craft 3.7, any plugin that provides mutations that extend craft\gql\base\ElementMutationResolver and calls $this->populateElementWithData($element, $arguments) must amend this function call to $this->populateElementWithData($element, $arguments, $resolveInfo), where $resolveInfo is the same variable that is passed to the resolver method as the last argument.

This will ensure that Craft traverses the arguments while keeping track of the argument GQL type definitions and will allow Craft to call all the nested value normalizers (if any)."

**Comment by humandevs** [comment]
This actually is not needed. Nested matrix fields inside a Super Table field can now be mutated via GraphQL, as long as you use the latest dev commit of Super Table (has not been packaged into a release yet). This can be closed as soon as the latest dev snapshot is released.

Here's an example mutation:
`mutation MyMutation($articleId: ID = "", $stepBuilderSortOrder: [QueryArgument] = "", $siteId: Int = 10, $stepBuilderId: ID = "", $blockSortOrder: [QueryArgument] = "", $textBlockId: ID = "", $textBlockParagraph: String = "") {
  save_articles_articleWithSteps_Entry(id: $articleId, topicArticleStepBuilder: {sortOrder: $stepBuilderSortOrder, blocks: {topicArticleStepBuilder_0: {id: $stepBuilderId, step: {sortOrder: $blockSortOrder, blocks: {textBlock: {id: $textBlockId, paragraph: $textBlockParagraph}}}}}}, siteId: $siteId) {
    id
    userTitle
    siteId
    topicArticleStepBuilder {
      ... on topicArticleStepBuilder_BlockType {
        id
        sortOrder
        step {
          ... on step_textBlock_BlockType {
            id
            paragraph
            title
            sortOrder
            typeHandle
          }
          ... on step_textPlusImageBlock_BlockType {
            id
            image {
              url
              title
              id
            }
            paragraph
            sortOrder
            title
            typeHandle
          }
          ... on step_largeImageBlock_BlockType {
            id
            uri
            image {
              id
              url
              title
            }
            sortOrder
            typeHandle
          }
          ... on step_calloutBlock_BlockType {
            id
            sortOrder
            typeHandle
            typeId
          }
          ... on step_videoBlock_BlockType {
            id
            title
            typeHandle
            uri
            youtubeVideo
            sortOrder
          }
        }
      }
    }
  }
}`

`{
  "articleId": "790",
  "stepBuilderSortOrder": ["791","794","796","798","800"],
  "siteId": 1,
  "stepBuilderId": "791",
  "blockSortOrder": ["861","793"],
  "textBlockId": "861",
  "textBlockParagraph": "Changed via GQL step"
}`

**Comment by engram-design** [comment]
Added the changes for the next release. To get this early, change your `verbb/super-table` requirement in `composer.json` to:

```
"require": {
  "verbb/super-table": "dev-craft-3 as 2.6.8",
  "...": "..."
}
```

Then run `composer update`. I know you're probably already running this, but leaving this here as a note to myself!

**Comment by reilnuud** [comment]
We're running into an issue where the mutation schema is using the supertable field's ID (an integer) as the name in GraphQL instead of the field's handle--so for us humandevs' solution looks more like this:

```graphql
[...]
topicArticleStepBuilder: {
      sortOrder: $stepBuilderSortOrder
      blocks: {
        topicArticleStepBuilder_0: {
          id: $stepBuilderId
          step: {
            sortOrder: $blockSortOrder
            blocks: {
              2: { id: $textBlockId, paragraph: $textBlockParagraph }
            }
          }
        }
      }
    }
 [...]
 ```

And using an integer (`2` in the example above) for a name throws all sorts of errors into the mix, since gql doesn't support it. Is there some setting we're missing or is this a bug? Using the actual block handle (`textBlock` in humandevs' example) returns a `Field \"textBlock\" not defined...`. Happy to open another issue if this is unrelated.

**Comment by reilnuud** [comment]
> We're running into an issue where the mutation schema is using the supertable field's ID (an integer) as the name in GraphQL instead of the field's handle--so for us humandevs' solution looks more like this:
> 
> ```graphql
> [...]
> topicArticleStepBuilder: {
>       sortOrder: $stepBuilderSortOrder
>       blocks: {
>         topicArticleStepBuilder_0: {
>           id: $stepBuilderId
>           step: {
>             sortOrder: $blockSortOrder
>             blocks: {
>               2: { id: $textBlockId, paragraph: $textBlockParagraph }
>             }
>           }
>         }
>       }
>     }
>  [...]
> ```
> 
> And using an integer (`2` in the example above) for a name throws all sorts of errors into the mix, since gql doesn't support it. Is there some setting we're missing or is this a bug? Using the actual block handle (`textBlock` in humandevs' example) returns a `Field \"textBlock\" not defined...`. Happy to open another issue if this is unrelated.

We were able to rename the field name in gql by resaving with a new name, clearing the gql cache, then changing it back, and cleaing the gql cache.

**Comment by humandevs** [comment]
> We're running into an issue where the mutation schema is using the supertable field's ID (an integer) as the name in GraphQL instead of the field's handle--so for us humandevs' solution looks more like this:
> 
> ```graphql
> [...]
> topicArticleStepBuilder: {
>       sortOrder: $stepBuilderSortOrder
>       blocks: {
>         topicArticleStepBuilder_0: {
>           id: $stepBuilderId
>           step: {
>             sortOrder: $blockSortOrder
>             blocks: {
>               2: { id: $textBlockId, paragraph: $textBlockParagraph }
>             }
>           }
>         }
>       }
>     }
>  [...]
> ```
> 
> And using an integer (`2` in the example above) for a name throws all sorts of errors into the mix, since gql doesn't support it. Is there some setting we're missing or is this a bug? Using the actual block handle (`textBlock` in humandevs' example) returns a `Field \"textBlock\" not defined...`. Happy to open another issue if this is unrelated.

Not sure if this is the issue at all, but just a note, CraftCMS GQL *queries* give you sortOrders as integers 1,2,3... but *mutations* require sortOrder as a list of actual block IDs. 

Also, the "textBlockId" in this case has to be one of the IDs listed in the sortOrder. Because of this we had to do a bunch of lookups to figure out what the sortOrder needed to be on mutation, by basically doing a full Entry query, and then filtering through that to lookup what block IDs existed and creating the list of block IDs to put in the mutation sortOrder. (It would be really nice if Craft offered sortOrderBlockIDs and sortOrderIndex fields to query, instead).

Also, the entry ID is missing from your example, I think this is required.

**Comment by reilnuud** [comment]
> > We're running into an issue where the mutation schema is using the supertable field's ID (an integer) as the name in GraphQL instead of the field's handle--so for us humandevs' solution looks more like this:
> > ```graphql
> > [...]
> > topicArticleStepBuilder: {
> >       sortOrder: $stepBuilderSortOrder
> >       blocks: {
> >         topicArticleStepBuilder_0: {
> >           id: $stepBuilderId
> >           step: {
> >             sortOrder: $blockSortOrder
> >             blocks: {
> >               2: { id: $textBlockId, paragraph: $textBlockParagraph }
> >             }
> >           }
> >         }
> >       }
> >     }
> >  [...]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And using an integer (`2` in the example above) for a name throws all sorts of errors into the mix, since gql doesn't support it. Is there some setting we're missing or is this a bug? Using the actual block handle (`textBlock` in humandevs' example) returns a `Field \"textBlock\" not defined...`. Happy to open another issue if this is unrelated.
> 
> Not sure if this is the issue at all, but just a note, CraftCMS GQL _queries_ give you sortOrders as integers 1,2,3... but _mutations_ require sortOrder as a list of actual block IDs.
> 
> Also, the "textBlockId" in this case has to be one of the IDs listed in the sortOrder. Because of this we had to do a bunch of lookups to figure out what the sortOrder needed to be on mutation, by basically doing a full Entry query, and then filtering through that to lookup what block IDs existed and creating the list of block IDs to put in the mutation sortOrder. (It would be really nice if Craft offered sortOrderBlockIDs and sortOrderIndex fields to query, instead).
> 
> Also, the entry ID is missing from your example, I think this is required.

> > We're running into an issue where the mutation schema is using the supertable field's ID (an integer) as the name in GraphQL instead of the field's handle--so for us humandevs' solution looks more like this:
> > ```graphql
> > [...]
> > topicArticleStepBuilder: {
> >       sortOrder: $stepBuilderSortOrder
> >       blocks: {
> >         topicArticleStepBuilder_0: {
> >           id: $stepBuilderId
> >           step: {
> >             sortOrder: $blockSortOrder
> >             blocks: {
> >               2: { id: $textBlockId, paragraph: $textBlockParagraph }
> >             }
> >           }
> >         }
> >       }
> >     }
> >  [...]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And using an integer (`2` in the example above) for a name throws all sorts of errors into the mix, since gql doesn't support it. Is there some setting we're missing or is this a bug? Using the actual block handle (`textBlock` in humandevs' example) returns a `Field \"textBlock\" not defined...`. Happy to open another issue if this is unrelated.
> 
> Not sure if this is the issue at all, but just a note, CraftCMS GQL _queries_ give you sortOrders as integers 1,2,3... but _mutations_ require sortOrder as a list of actual block IDs.
> 
> Also, the "textBlockId" in this case has to be one of the IDs listed in the sortOrder. Because of this we had to do a bunch of lookups to figure out what the sortOrder needed to be on mutation, by basically doing a full Entry query, and then filtering through that to lookup what block IDs existed and creating the list of block IDs to put in the mutation sortOrder. (It would be really nice if Craft offered sortOrderBlockIDs and sortOrderIndex fields to query, instead).
> 
> Also, the entry ID is missing from your example, I think this is required.

Yeah, it was the ID of the supertable field type that was getting grabbed by the mutation schema--we're now getting something more like `fieldTypeHandle_0` instead of simply an integer after renaming the field handle.

**Comment by engram-design** [comment]
Should be fixed in [2.7.0](https://github.com/verbb/super-table/releases/tag/2.7.0)
